### ISSUE TITLE
Rails association between Persons and Users

### ISSUE THREAD

**Issue by jonatanklosko** [cubing/worldcubeassociation.org#619]
Hi, I'm in progress of porting `change_person.php` admin script to Rails and I've stumbled upon a huge issue that I've no idea how to fix.
So the problem is that we have the composite primary key on Persons table (id, subId). Despite of that in our Rails code we have [only id set as a primary key](https://github.com/cubing/worldcubeassociation.org/blob/master/WcaOnRails/app/models/person.rb#L3). I guess, this hasn't been a problem so far because the script I mentioned has been handling all the updates to the Persons table.
Some real example demonstrating the problem:
```ruby
# Let's assume that we have two persons (p1 and p2) with the same id (2016MYID01)
# and different subIds (1 and 2)
p1.update_attributes name: "Person One"
# This produces the following SQL query
UPDATE `Persons` SET `name` = 'Person One' WHERE `Persons`.`id` = '2016MYID01'
# As you can see since we have id set as the primary key in the model code,
# this update generates query independent from the subId of this person
# This causes both p1 and p2 to be updated with the new name
```
Briefly the things we need to change are:
- Add composite primary key to the Rails code
- Make the Person - User association to use subId equal 1 (so the user is linked with the current person record)

I've already found a [gem](https://github.com/composite-primary-keys/composite_primary_keys) but this doesn't seem to support the second point above. (Also it has an issue with tables having `id` column.)

@jfly @timhabermaas @larspetrus do you have any ideas?

**Comment by larspetrus** [comment]
Are there actually separate people with the same ID in the DB?

On Sun, May 15, 2016 at 10:21 AM, Jonatan Kłosko <notifications@github.com>
wrote:

> Hi, I'm in progress of porting change_person.php admin script to Rails
> and I've stumbled upon a huge issue that I've no idea how to fix.
> So the problem is that we have the composite primary key on Persons table
> (id, subId). Despite of that in our Rails code we have only id set as a
> primary key
> <https://github.com/cubing/worldcubeassociation.org/blob/master/WcaOnRails/app/models/person.rb#L3>.
> I guess, this hasn't been a problem so far because the script I mentioned
> has been handling all the updates to the Persons table.
> Some real example demonstrating the problem:
>
> # Let's assume that we have two persons (p1 and p2) with the same id (2016MYID01)# and different subIds (1 and 2)
> p1.update_attributes name: "Person One"# This produces the following SQL queryUPDATE `Persons` SET `name` = 'Person One' WHERE `Persons`.`id` = '2016MYID01'# As you can see since we have id set as the primary key in the model code,# this update generates query independent from the subId of this person# This causes both p1 and p2 to be updated with the new name
>
> Briefly the things we need to change are:
>
>    - Add composite primary key to the Rails code
>    - Make the Person - User association to use subId equal 1 (so the user
>    is linked with the current person record)
>
> I've already found a gem
> <https://github.com/composite-primary-keys/composite_primary_keys> but
> this doesn't seem to support the second point above. (Also it has an issue
> with tables having id column.)
>
> @jfly <https://github.com/jfly> @timhabermaas
> <https://github.com/timhabermaas> @larspetrus
> <https://github.com/larspetrus> do you have any ideas?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/cubing/worldcubeassociation.org/issues/619>
>



-- 
Procrastinators - the Leaders of Tomorrow

**Comment by larspetrus** [comment]
Trying to contort Rails into using a composite primary key is probably
somewhere between painful and impossible.

My instinct is to add a *rails_id* column that is simply *"#{id}_#{subId}"*.
So the person in your example would have rails_id *2016MYID01_2*.

If you keep this column in sync, things should just work, and while that's
non trivial, I don't think it's hard. Crucially, it's very easy to see when
you've failed.

On Sun, May 15, 2016 at 10:30 AM, Lars Petrus <lars@lar5.com> wrote:

> Are there actually separate people with the same ID in the DB?
>
> On Sun, May 15, 2016 at 10:21 AM, Jonatan Kłosko <notifications@github.com
> > wrote:
>
>> Hi, I'm in progress of porting change_person.php admin script to Rails
>> and I've stumbled upon a huge issue that I've no idea how to fix.
>> So the problem is that we have the composite primary key on Persons table
>> (id, subId). Despite of that in our Rails code we have only id set as a
>> primary key
>> <https://github.com/cubing/worldcubeassociation.org/blob/master/WcaOnRails/app/models/person.rb#L3>.
>> I guess, this hasn't been a problem so far because the script I mentioned
>> has been handling all the updates to the Persons table.
>> Some real example demonstrating the problem:
>>
>> # Let's assume that we have two persons (p1 and p2) with the same id (2016MYID01)# and different subIds (1 and 2)
>> p1.update_attributes name: "Person One"# This produces the following SQL queryUPDATE `Persons` SET `name` = 'Person One' WHERE `Persons`.`id` = '2016MYID01'# As you can see since we have id set as the primary key in the model code,# this update generates query independent from the subId of this person# This causes both p1 and p2 to be updated with the new name
>>
>> Briefly the things we need to change are:
>>
>>    - Add composite primary key to the Rails code
>>    - Make the Person - User association to use subId equal 1 (so the
>>    user is linked with the current person record)
>>
>> I've already found a gem
>> <https://github.com/composite-primary-keys/composite_primary_keys> but
>> this doesn't seem to support the second point above. (Also it has an issue
>> with tables having id column.)
>>
>> @jfly <https://github.com/jfly> @timhabermaas
>> <https://github.com/timhabermaas> @larspetrus
>> <https://github.com/larspetrus> do you have any ideas?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly or view it on GitHub
>> <https://github.com/cubing/worldcubeassociation.org/issues/619>
>>
>
>
>
> --
> Procrastinators - the Leaders of Tomorrow
>



-- 
Procrastinators - the Leaders of Tomorrow

**Comment by jonatanklosko** [comment]
Hmm, that seems reasonable but if it means that we need to add another column in the users like `person_rails_id` it becomes really messy and even more confusing for people not familiar with this.

Just thinking out loud: I consider this `subId` thing to be a bit cumbersome because it's almost never used whereas we need to use `WHERE subId = 1` in the PHP code. I'm not even sure if it's necessary since all the informations that may change and matter for results (name, country) are already in the results table.

**Comment by jonatanklosko** [comment]
Fixed by #645.
